src/placeos-rest-api/controllers/metadata.cr

```diff
@@ -56,22 +56,37 @@ module PlaceOS::Api
     ###############################################################################################
 
     def index
+      # Construct the queries from the URI parameters
       queries = query_params.compact_map do |key, value|
         Model::Metadata::Query.from_param?(key, value.presence)
       end
 
-      results = Model::Metadata.query(queries)
-      total = results.size
-
-      # TODO: Add pagination to `Metadata#index`
-      # range_start = offset
-      # range_end = (results.size || 0) + range_start
+      # TODO: Use destructure after `spider-gazelle/promise` is fixed
+      data = Promise.all(
+        Promise.defer { Model::Metadata.query(queries, offset, limit) },
+        Promise.defer { Model::Metadata.query_count(queries) },
+      ).get
+
+      total = 0
+      results = Array(Model::Metadata).new(initial_capacity: 0)
+      data.each do |element|
+        case element
+        in Int32                  then total = element
+        in Array(Model::Metadata) then results = element
+        end
+      end
 
-      range_start = 0
-      range_end = total
+      range_end = results.size + offset
 
       response.headers["X-Total-Count"] = total.to_s
-      response.headers["Content-Range"] = "metadata #{range_start}-#{range_end}/#{total}"
+      response.headers["Content-Range"] = "metadata #{offset}-#{range_end}/#{total}"
+
+      # Set link if further results
+      if range_end < total
+        query_params["offset"] = (range_end + 1).to_s
+        query_params["limit"] = limit.to_s
+        response.headers["Link"] = %(<#{base_route}?#{query_params}>; rel="next")
```

Comment:
This is the only line I'm unsure about without knowing where `query_params` is defined. I'm assuming it's not a simple `Hash(String,Int32 | String)` otherwise this won't produce a valid query string. If that's the case just ignore this comment and everything else appears fine.

Reply:
It's one of these https://crystal-lang.org/api/1.4.1/URI/Params.html
The `to_s` turns it into a query string